llms/mlx_lm/utils.py

```diff
@@ -154,13 +189,16 @@ def generate(
     skip = 0
     REPLACEMENT_CHAR = "\ufffd"
 
-    for (token, prob), n in zip(generate_step(prompt, model, temp), range(max_tokens)):
-        if token == tokenizer.eos_token_id:
+    for (token, prob), n in zip(
```

Comment:
Personally, I feel that we should accept a parameter that allows the user to decide the number of tokens to apply for repetition_penalty. We should also maintain a separate list to use as repetition_penalty context instead of using the existing tokens list. This could cause issues when applying repetition_penalty to very long generated token lists.

Reply:
I actually thought about this, what are your thoughts about using a queue instead of a list to maintain the number of tokens to apply for repetition_penalty?

We can initialize the queue with a maxlen of number of tokens to apply for repetition penalty. The tokens in the queue would be automatically popping from the front as new tokens are added to the queue. Kind of like a fixed sliding window over the tokens.

Possible implementation:

```python
def generate_step(prompt: mx.array, model: nn.Module, temp: float, repetition_penalty: float = 1.0,
    repetition_context: Optional[Any] = None # previously generated_tokens
) -> Generator[Tuple[mx.array, mx.array], None, None]:
    
    # def sample
    
    y = prompt
    cache = None

    while True:
        logits, cache = model(y[:, None], cache=cache)
        logits = logits.squeeze(1)

        if repetition_context is not None:
            logits = apply_penalty(logits, repetition_context, repetition_penalty)

        y, prob = sample(logits)
        yield y, prob
        
def generate(model: nn.Module, tokenizer: PreTrainedTokenizer, prompt: str, temp: Optional[float], 
repetition_penalty: Optional[float], repetition_context_size: Optional[int], 
max_tokens: int = 100, verbose: bool = False, formatter: Callable = None,
) -> str:
 
    prompt_tokens = mx.array(tokenizer.encode(prompt))
    repetition_context = deque(maxlen=repetition_context_size)
    repetition_context.extend(tokenizer.encode(prompt))

    for (token, prob), n in zip(
        generate_step(prompt=prompt_tokens, model= model, 
                      temp=temp, repetition_penalty=repetition_penalty, 
                      repetition_context=repetition_context), range(max_tokens)):
                      
        if token[0] == tokenizer.eos_token_id:
            break
        if n == 0:
            prompt_time = time.perf_counter() - tic
            tic = time.perf_counter()
        tokens.append(token[0].item())

        if verbose:
           .....

        repetition_context.append(token[0].item())  # Update repetition context after each token

    tokens = tokenizer.decode(tokens).replace(REPLACEMENT_CHAR, "")
    
```

Reply:
As long as it is simple and easy to understand, I personally don't have a preference either way. However, it would be great to maintain the same coding style that mlx-lm currently uses.